Add child logger support and module handling.

davidlehn opened the pull request:
- Add `.child(meta)` method to loggers that has common metadata.
- The special `module` key is used to prefix messages with `[module]`
  and then removed from the metadata. These adjustments may be removed in
  the future but are necessary with the current lower level logging
  mechanisms.

This is used like:

``` js
var logger = bedrock.loggers.get('app').child({module: 'bedrock-foobar'});
logger.info('ima log');
```

and results in:

```
2016-10-05T18:24:53.927Z - info: [bedrock-foobar] ima log workerPid=9611
```

I'm not so thrilled about how this patch works.  It adjusts the msg text and removes the module meta if present. That makes it harder to use logging handlers that want all the raw semantic data. I'm not sure if there's an easy way to get console and file loggers alone to do this manipulation. The module metadata removal is to avoid it being printed out for every log. It may be ok to just use this for now until a better solution comes about. It's not too terrible that the module data is in the message but could potentially stop some automatic module queries. Thoughts?

dlongley commented:
So we want this API and functionality, but it's probably not quite the right way to do it. We probably want to override file/console loggers to handle the formatting so that the semantic data (the module name) stays in the meta data. That being said, let's proceed with this for now but please add a note to the code itself with something about this "potentially better" way.

davidlehn commented:
I haven't merged this yet since I was exploring improvements and unsure if the API would need to be changed.  I was trying to pass through the raw data to loggers and override them to pretty print as needed.  But there are complications and it's been sitting around for a while.

mattcollier commented on README.md:
```diff
@@ -693,6 +693,12 @@ logger can be retrieved via `bedrock.loggers.get('app')`. A call to
+be added to child log output. A special `module` meta name can optionally be
+used for pretty output. A shortcut for creating named module loggers is
+`bedrock.loggers.get('app').child('name')`.
+
```
could we add an example of the log output you get with this?

is it `error: [module_name] My Message` ?

davidlehn replied:
Yeah, looks like that. I'll add an example.

davidlehn commented:
I tried to improve this a bit.  `child(name)` now a shortcut for `child({module: name})`.  Also moved the message module prefix mangling into transport subclasses.  So more semantic usage possible if you want it.  Also added config option to turn on/off the module prefix per category.  Current patch has module prefix disabled for file output.  Mostly for testing.  It will show up as `{time} - info: mumble mumble module=bedrock-foobar` in files but `{time} - info: [bedrock-foobar] mumble mumble` on the console.  Easily adjustable.

This can replace all places where we currently have added `[module-name]` logging prefixes per message.  Also could be nice to expand usage and get most of the log messages to have module namespaces.  Might be able to replace redundancy in some logging text if having the module name shown makes it obvious you don't need that explanation in the main text.

Also could be used like in [bunyan](https://github.com/trentm/node-bunyan) examples to add per request loggers and such if we need that somewhere.
